op-e2e/actions/upgrades/interop_fork_test.go

```diff
@@ -0,0 +1,138 @@
+	rollupConfigB := is.Out.L2s[actors.ChainB.ChainID.String()].RollupCfg
+
+	currentBlockHeaderA := actors.ChainA.SequencerEngine.L2Chain().CurrentBlock()
+	currentBlockHeaderB := actors.ChainB.SequencerEngine.L2Chain().CurrentBlock()
+	require.False(t, rollupConfigA.IsInterop(currentBlockHeaderA.Time),
+		"Interop should not be active at the first L1 inclusion block in chain A")
+	require.False(t, rollupConfigB.IsInterop(currentBlockHeaderB.Time),
+		"Interop should not be active at the first L1 inclusion block in chain B")
+
+	// Build a few L2 blocks. We only need the L1 inclusion to advance past Interop and Interop
+	// shouldn't activate with L2 time.
+	actors.ChainA.Sequencer.ActBuildL2ToInterop(t)
+	actors.ChainB.Sequencer.ActBuildL2ToInterop(t)
+
+	////////////////////////////
+	// Post-upgrade Block Production
+	////////////////////////////
+
+	currentBlockHeaderA = actors.ChainA.SequencerEngine.L2Chain().CurrentBlock()
+	currentBlockHeaderB = actors.ChainB.SequencerEngine.L2Chain().CurrentBlock()
+	require.Equal(t, currentBlockHeaderA.Number.Uint64(), currentBlockHeaderB.Number.Uint64())
+	currentBlockNumber := currentBlockHeaderA.Number
+
+	require.True(t, rollupConfigA.IsInteropActivationBlock(currentBlockHeaderA.Time),
+		"Interop should be active at the first L1 inclusion block in chain A")
+	require.True(t, rollupConfigB.IsInteropActivationBlock(currentBlockHeaderB.Time),
+		"Interop should be active at the first L1 inclusion block in chain B")
+
+	currentBlock := actors.ChainA.SequencerEngine.L2Chain().GetBlockByHash(currentBlockHeaderA.Hash())
+	activationBlockTxs := currentBlock.Transactions()
+
+	require.Len(t, activationBlockTxs, 5) // 4 upgrade txs + 1 system deposit tx
+	upgradeTransactions := activationBlockTxs[1:]
+	upgradeTransactionBytes := make([]hexutil.Bytes, len(upgradeTransactions))
+	for i, tx := range upgradeTransactions {
+		txBytes, err := tx.MarshalBinary()
+		require.NoError(t, err)
+		upgradeTransactionBytes[i] = txBytes
+	}
+
+	expectedUpgradeTransactions, err := derive.InteropNetworkUpgradeTransactions()
+	require.NoError(t, err)
+
+	require.ElementsMatch(t, upgradeTransactionBytes, expectedUpgradeTransactions)
+
+	// check CrossL2Inbox deployed correctly
+	proxyImplGetterFunc := w3.MustNewFunc(`implementation()`, `address`)
+	proxyImplGetterSelector, err := proxyImplGetterFunc.EncodeArgs()
+	require.NoError(t, err)
+
+	code, err := actors.ChainA.SequencerEngine.EthClient().CodeAt(t.Ctx(), derive.CrossL2InboxAddress, currentBlockNumber)
+	require.NoError(t, err)
+	require.NotEmpty(t, code, "CrossL2Inbox contract should be deployed")
+
+	implementationBytes, err := actors.ChainA.SequencerEngine.EthClient().CallContract(t.Ctx(), ethereum.CallMsg{
+		To:   &predeploys.CrossL2InboxAddr,
+		Data: proxyImplGetterSelector,
+	}, currentBlockNumber)
+	require.NoError(t, err)
+	var implAddr common.Address
+	err = proxyImplGetterFunc.DecodeReturns(implementationBytes, &implAddr)
+	require.NoError(t, err)
```

Comment:
Would be great to extract the code that calls `implementation()` on a proxy address with an ethclient (ideally use `interface{ CallContract(...) }` into a helper. Would help readability and reduce code in this (and future similar) tests.

Reply:
Fixed in c6fd3b743583a4bdbc626c27c8e5b1a8af210d24.